core/src/main/java/com/datastax/oss/driver/api/core/paging/OffsetPager.java

```diff
@@ -0,0 +1,312 @@
+ *
+ * In concrete terms, this class <em>does</em> start from the beginning and skip rows until it
+ * reaches the desired page, every time. You need to re-execute the query in order to start with a
+ * fresh result set for each invocation:
+ *
+ * <pre>
+ * String query = "SELECT ...";
+ * OffsetPager pager = new OffsetPager(20);
+ *
+ * ResultSet rs = session.execute(query);
+ * // Iterate through rows 1-20 and discard them, then return rows 21-40
+ * OffsetPager.Page&lt;Row&gt; page2 = pager.getPage(rs, 2);
+ *
+ * rs = session.execute(query);
+ * // Iterate through rows 1-80 and discard them, then return rows 81-100
+ * OffsetPager.Page&lt;Row&gt; page5 = pager.getPage(rs, 5);
+ * </pre>
+ *
+ * This is fine for the values typically encountered in real-world applications: for example, if the
+ * page size is 25 and users never go past page 10, the worst case is only 250 rows, which is a very
+ * small result set.
+ *
+ * <p>On the other hand, if you want to display 10,000 elements per page and allow the user to query
+ * page 7,384, this class is not going to work. You should actively defend against such aberrant
+ * values, see the next paragraph.
+ *
+ * <h3>Establishing application-level guardrails</h3>
+ *
+ * If you use this class, we strongly recommend that you implement hard limits in your application
+ * code. If the page number is exposed to the user (for example if it is passed as a URL parameter),
+ * make sure it is properly validated and enforce a maximum, so that an attacker can't inject a
+ * large value that could potentially fetch millions of rows.
+ *
+ * <h3>Relation with protocol-level paging</h3>
+ *
+ * There is also something called "paging" at the protocol level (controlled by the config option
+ * {@code basic.request.page-size}). It simply means that the driver might receive the results in
+ * multiple "chunks"; this happens under the hood and is completely transparent for users of this
+ * class. You don't need to set the protocol page size to the same value as the logical page size.
+ */
+@ThreadSafe
+public class OffsetPager {
+
+  /** A page returned as the result of an offset query. */
+  public interface Page<ElementT> {
+
+    /** The elements in the page. */
+    @NonNull
+    List<ElementT> getElements();
+
+    /**
+     * The page number (1 for the first page, 2 for the second page, etc).
+     *
+     * <p>Note that it may be different than the requested page number: if the result set is too
+     * short, {@link OffsetPager#getPage(PagingIterable, int)} returns the last page instead, and
+     * this method will return the number of that last page.
+     */
+    int getPageNumber();
+
+    /** Whether this is the last page in the result set. */
+    boolean isLast();
+  }
+
+  private final int pageSize;
+
+  /**
+   * Creates a new instance.
+   *
+   * @param pageSize the number of elements per page. Must be greater than or equal to 1.
+   */
+  public OffsetPager(int pageSize) {
```

Comment:
While writing some examples I realized it's more convenient to pass the page size once through the constructor. Also, it prevents accidentally inverting page size and page number.

The only downside is if you use different page sizes across your application, but this class holds no state so creating multiple instances does not add a lot of overhead.